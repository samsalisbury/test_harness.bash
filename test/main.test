#!/usr/bin/env bash

# shellcheck disable=SC1090
source "${TESTHARNESS:-../test_harness.bash}"

# ABS_TESTHARNESS is use from tests in arbitrary work dirs.
export ABS_TESTHARNESS="$PWD/../test_harness.bash"

add_test_harness() { cp "$ABS_TESTHARNESS" ./; }

make_test_file() { local NAME="$1" CONTENT="$2"
  mkdir -p "$(dirname "$NAME")"
  printf "#!/usr/bin/env bash\n\nset -euo pipefail\n\n" > "$NAME"
  printf "source test_harness.bash\n\n" >> "$NAME"
  echo "$CONTENT" >> "$NAME"
  chmod +x "$NAME"
}

(
  begin_test pass/empty
  
  add_test_harness
  make_test_file "the.test" "
  (
    begin_test $TEST_ID
  )
  "

  run env VERBOSE=YES ./the.test || error "empty test failed"
)

(
  begin_test pass/true
  
  add_test_harness
  make_test_file "the.test" "
  (
    begin_test $TEST_ID
    true
  )
  "

  run env VERBOSE=YES ./the.test || error "bare true test failed"
)

(
  begin_test pass/log
  
  add_test_harness
  make_test_file "the.test" "
  (
    begin_test $TEST_ID
    log 'Hello %s' World
  )
  "

  run env VERBOSE=YES ./the.test || error "log failed test"
)

(
  begin_test fail/false
  
  add_test_harness
  make_test_file "the.test" "
  (
    begin_test $TEST_ID
    false
  )
  "

  if run env VERBOSE=YES ./the.test; then
    error "bare false did not fail test"
  fi
)

(
  begin_test fail/error
  
  add_test_harness
  make_test_file "the.test" "
  (
    begin_test $TEST_ID
    error 'this is an error'
  )
  "

  if run env VERBOSE=YES ./the.test; then
    error "error did not fail test"
  fi
)

(
  begin_test fail/fatal
  
  add_test_harness
  make_test_file "the.test" "
  (
    begin_test $TEST_ID
    fatal 'this is a fatal error'
    log 'you should never see this'
  )
  "

  if run env VERBOSE=YES ./the.test; then
    error "fatal did not fail test"
  fi

  [[ "$STDOUT" = *"this is a fatal error"* ]] || error "fatal error not reported"
  [[ "$STDOUT" = *"you should never see this"* ]] && error "fatal error did not end test"
  [[ -z "$STDERR" ]] || error "stderr not empty, contains %s" "$STDERR"
  [[ "$EXIT_CODE" = 1 ]] || error "got exit code %s; want %s" "$EXIT_CODE" 1
)
