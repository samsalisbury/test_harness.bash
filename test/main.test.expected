=== RUN   main/Test/pass/empty
--- PASS: main/Test/pass/empty
    $ ./the.test -v -notime -run .
    === RUN   the/TestEmpty
    --- PASS: the/TestEmpty
    PASS
    ok        the
    $ ./testing.bash -v -notime -run . ./...
    === RUN   the/TestEmpty
    --- PASS: the/TestEmpty
    PASS
    ok        the
=== RUN   main/Test/pass/true
--- PASS: main/Test/pass/true
    $ ./the.test -v -notime -run .
    === RUN   the/TestTrue
    --- PASS: the/TestTrue
    PASS
    ok        the
    $ ./testing.bash -v -notime -run . ./...
    === RUN   the/TestTrue
    --- PASS: the/TestTrue
    PASS
    ok        the
=== RUN   main/Test/pass/log
--- PASS: main/Test/pass/log
    $ ./the.test -v -notime -run .
    === RUN   the/TestLog
    --- PASS: the/TestLog
        the.test:8: Hello World
    PASS
    ok        the
    $ ./testing.bash -v -notime -run . ./...
    === RUN   the/TestLog
    --- PASS: the/TestLog
        the.test:8: Hello World
    PASS
    ok        the
=== RUN   main/Test/pass/logformat
--- PASS: main/Test/pass/logformat
    $ ./the.test -v -notime -run .
    === RUN   the/TestLog
    --- PASS: the/TestLog
        the.test:8: Hello, cruel World
    PASS
    ok        the
    $ ./testing.bash -v -notime -run . ./...
    === RUN   the/TestLog
    --- PASS: the/TestLog
        the.test:8: Hello, cruel World
    PASS
    ok        the
=== RUN   main/Test/fail/false
--- PASS: main/Test/fail/false
    $ ./the.test -v -notime -run .
    === RUN   the/TestFalse
    --- FAIL: the/TestFalse
        the.test:8: Command failed with exit code 1: false
    FAIL
    fail      the
    $ ./testing.bash -v -notime -run . ./...
    === RUN   the/TestFalse
    --- FAIL: the/TestFalse
        the.test:8: Command failed with exit code 1: false
    FAIL
    fail      the
=== RUN   main/Test/fail/exit1_from_test
--- PASS: main/Test/fail/exit1_from_test
    $ ./the.test -v -notime -run .
    === RUN   the/TestExit1
    --- FAIL: the/TestExit1
        Test body failed with exit code 1
    FAIL
    fail      the
    $ ./testing.bash -v -notime -run . ./...
    === RUN   the/TestExit1
    --- FAIL: the/TestExit1
        Test body failed with exit code 1
    FAIL
    fail      the
=== RUN   main/Test/fail/exit2_before_test
--- PASS: main/Test/fail/exit2_before_test
    $ ./the.test -v -notime -run .
    $ ./testing.bash -v -notime -run . ./...
=== RUN   main/Test/fail/exit2_after_test
--- PASS: main/Test/fail/exit2_after_test
    $ ./the.test -v -notime -run .
    $ ./testing.bash -v -notime -run . ./...
=== RUN   main/Test/fail/error
--- PASS: main/Test/fail/error
    $ ./the.test -v -notime -run .
    === RUN   the/TestError
    --- FAIL: the/TestError
        the.test:9: this is an error
        the.test:10: but the test continues and fails
    FAIL
    fail      the
    $ ./testing.bash -v -notime -run . ./...
    === RUN   the/TestError
    --- FAIL: the/TestError
        the.test:9: this is an error
        the.test:10: but the test continues and fails
    FAIL
    fail      the
=== RUN   main/Test/fail/errorformat
--- PASS: main/Test/fail/errorformat
    $ ./the.test -v -notime -run .
    === RUN   the/TestError
    --- FAIL: the/TestError
        the.test:9: this is a formatted, terrible error
        the.test:10: but the test continues and fails
    FAIL
    fail      the
    $ ./testing.bash -v -notime -run . ./...
    === RUN   the/TestError
    --- FAIL: the/TestError
        the.test:9: this is a formatted, terrible error
        the.test:10: but the test continues and fails
    FAIL
    fail      the
=== RUN   main/Test/fail/fatal
--- PASS: main/Test/fail/fatal
    $ ./the.test -v -notime -run .
    === RUN   the/TestFatal
    --- FAIL: the/TestFatal
        the.test:9: this is a fatal error
    FAIL
    fail      the
    $ ./testing.bash -v -notime -run . ./...
    === RUN   the/TestFatal
    --- FAIL: the/TestFatal
        the.test:9: this is a fatal error
    FAIL
    fail      the
=== RUN   main/Test/fail/fatalformat
--- PASS: main/Test/fail/fatalformat
    $ ./the.test -v -notime -run .
    === RUN   the/TestFatal
    --- FAIL: the/TestFatal
        the.test:9: this is a formatted fatal error
    FAIL
    fail      the
    $ ./testing.bash -v -notime -run . ./...
    === RUN   the/TestFatal
    --- FAIL: the/TestFatal
        the.test:9: this is a formatted fatal error
    FAIL
    fail      the
PASS
ok        main
